### PR TITLE
Add WarningsNotAsErrors option

### LINKED ISSUE #3395: WarningsNotAsErrors tag in project file is not implemented
Scenario: I want to treat all warnings as errors EXCEPT some specific ones. This is possible in when calling F# compiler directly (https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/compiler-options), but there is no way to specify this option in fsproj file.

Note that it is possible in C#: https://stackoverflow.com/a/468166/1108916

I found that there is `WarningsAsErrors` setting, but `WarningsNotAsErrors` is missing: https://github.com/Microsoft/visualfsharp/blob/f6049d54d464874559765ced35276978656a5e52/src/fsharp/FSharp.Build/Fsc.fsi#L62 this is probably where it should be implemented.

The option is also missing from the VS UI.

Apologies if this is not a correct place to report this.

### PR DESCRIPTION
**Pull request by kasper3**
Fixes #3395

### CODE REVIEW

File: fcs/FSharp.Compiler.Service.ProjectCrackerTool/ProjectCrackerTool.fs [hunk 1]
```diff
@@ -274,6 +274,7 @@ module internal ProjectCrackerTool =
       let defines = split (getProp "DefineConstants") [| ';'; ','; ' ' |]
       let nowarn = split (getProp "NoWarn") [| ';'; ','; ' ' |]
       let warningsAsError = split (getProp "WarningsAsErrors") [| ';'; ','; ' ' |]
+      let warningsAsError = split (getProp "WarningsNotAsErrors") [| ';'; ','; ' ' |]
```
Comment (cartermp) [inline fcs/FSharp.Compiler.Service.ProjectCrackerTool/ProjectCrackerTool.fs]: Should this be `warningsNotAsError`?

Comment (kasper3) [reply fcs/FSharp.Compiler.Service.ProjectCrackerTool/ProjectCrackerTool.fs]: Fixed :)

**Comment by kasper3** [comment]
Windows_NT Release_ci_part1-4 Build are showing errors like:

```
 building lookahead table...        time: 00:00:00.0068851
  building action table...        shift/reduce error at state 4 on terminal RPAREN between {[explicit nonassoc 9994] shift(24)} and {noprec reduce(Recover:'error')} - assuming the former because we prefer shift when unable to compare precedences
          shift/reduce error at state 14 on terminal error between {noprec shift(3)} and {[explicit left 9998] reduce(Expr:'OP_NOT' Expr)} - assuming the former because we prefer shift when unable to compare precedences
          shift/reduce error at state 15 on terminal error between {noprec shift(3)} and {[explicit left 9997] reduce(Expr:Expr 'OP_AND' Expr)} - assuming the former because we prefer shift when unable to compare precedences
          shift/reduce error at state 16 on terminal error between {noprec shift(3)} and {[explicit left 9996] reduce(Expr:Expr 'OP_OR' Expr)} - assuming the former because we prefer shift when unable to compare precedences
          reduce/reduce error at state 25 on terminal OP_AND between {[explicit nonassoc 9999] reduce(Expr:'LPAREN' Expr Recover)} and {noprec reduce(Expr:Expr Recover)} - assuming the former because we prefer the rule earlier in the file
          reduce/reduce error at state 25 on terminal OP_OR between {[explicit nonassoc 9999] reduce(Expr:'LPAREN' Expr Recover)} and {noprec reduce(Expr:Expr Recover)} - assuming the former because we prefer the rule earlier in the file
```

are these related to my changes?

File: fcs/FSharp.Compiler.Service.ProjectCrackerTool/ProjectCrackerTool.fs [hunk 2]
```diff
@@ -336,6 +337,8 @@ module internal ProjectCrackerTool =
                   if not (String.IsNullOrWhiteSpace nw) then yield "--nowarn:" + nw
               for nw in warningsAsError do
                   if not (String.IsNullOrWhiteSpace nw) then yield "--warnaserror:" + nw
+              for nw in warningsNotAsError do
+                  if not (String.IsNullOrWhiteSpace nw) then yield "--warnaswarn:" + nw
```
Comment (saul) [inline fcs/FSharp.Compiler.Service.ProjectCrackerTool/ProjectCrackerTool.fs]: This argument already exists as `--warnaserror-:`

Comment (kasper3) [reply fcs/FSharp.Compiler.Service.ProjectCrackerTool/ProjectCrackerTool.fs]: @saul, thanks! should I revert the changes in `src/fsharp/CompileOptions.fs` as well?

File: src/fsharp/FSharp.Build/Fsc.fs [hunk 3]
```diff
@@ -195,6 +196,11 @@ type public Fsc () as this =
 
         builder.AppendSwitchIfNotNull("--warnaserror:", warningsAsErrorsArray, ",")
 
+        // WarningsNotAsErrors
+        match warningsNotAsErrors with
+        | null -> ()
+        | _ -> builder.AppendSwitchIfNotNull("--warnaserror-:", warningsNotAsErrors.Split([|' '; ';'; ','|], StringSplitOptions.RemoveEmptyEntries), ",")
```
Comment (saul) [inline src/fsharp/FSharp.Build/Fsc.fs]: Won't we get a NullReferenceException if \<WarningsNotAsErrors\> isn't set?

Comment (kasper3) [reply src/fsharp/FSharp.Build/Fsc.fs]: We are not checking null for any other property and default value is null for each. This pattern is used few lines above (see https://github.com/Microsoft/visualfsharp/pull/4562/files#diff-bcb4028a918d13bbab2d0c1c15c97e6aR178, and default value of `disabledWarnings` is set to `null` at the top)

Comment (saul) [reply src/fsharp/FSharp.Build/Fsc.fs]: Sorry I'm blind - I didn't see the fact that we were in a match where we explicitly check for null

**Comment by saul** [comment]
Am I not right in saying that \<WarningsNotAsErrors\> values are global to the project? e.g. you'd set it to something like:

```
<WarningsNotAsErrors>$(WarningsNotAsErrors);NU1603;FS0040</WarningsNotAsErrors>
```

Is this going to work? Aren't we going to try to pass the NU1603 to the compiler here, which will cause an error? Do we need to strip the FS off?

**Comment by dsyme** [comment]
Please add a test, e.g. the FscTask tests in the unit tests for FSharp.Build.dll

**Comment by kasper3** [comment]
@dsyme do you have example how the msbuild tasks are unit tested, are your tests invoking msbuild System.Diagnostics.Process.Start?

**Comment by dsyme** [comment]
It looks like the tests are actually in 

    vsintegration\tests\UnitTests\Tests.Build.fs

I think it should be easy to copy and adapt one of those tests?

**Comment by pmbanka** [comment]
@kasper3 one thing I noticed is that you also probably need to update the signatures file (Fsc.fsi):

https://github.com/Microsoft/visualfsharp/blob/f6049d54d464874559765ced35276978656a5e52/src/fsharp/FSharp.Build/Fsc.fsi#L62

**Comment by kasper3** [comment]
Thanks for the feedback, I have added a test case.

**Comment by kasper3** [comment]
@dsyme, could you take another look? :)

**Comment by cartermp** [comment]
Ping for re-review @dsyme

**Comment by KevinRansom** [comment]
@kasper3 thanks for this

**Pull request merged**